streams/src/test/java/org/apache/kafka/streams/integration/KStreamRepartitionIntegrationTest.java

```diff
@@ -449,6 +449,85 @@ public void shouldPerformSelectKeyWithRepartitionOperation() throws Exception {
         assertEquals(1, countOccurrencesInTopology(topology, "Sink: .*-repartition.*"));
     }
 
+    @Test
+    public void shouldNotRepartitionWithMarkAsPartitionedFollowingSelectKey() throws Exception {
+        final long timestamp = System.currentTimeMillis();
+
+        sendEvents(
+                timestamp,
+                Arrays.asList(
+                        new KeyValue<>(1, "10"),
+                        new KeyValue<>(2, "20")
+                )
+        );
+
+        final StreamsBuilder builder = new StreamsBuilder();
+
+        builder.stream(inputTopic, Consumed.with(Serdes.Integer(), Serdes.String()))
+                .selectKey((key, value) -> Integer.valueOf(value))
+                .markAsPartitioned()
+                .groupByKey()
+                .count()
+                .toStream()
+                .to(outputTopic);
+
+
+        startStreams(builder);
+
+        validateReceivedMessages(
+                new IntegerDeserializer(),
+                new LongDeserializer(),
+                Arrays.asList(
+                        new KeyValue<>(10, 1L),
+                        new KeyValue<>(20, 1L)
+                )
+        );
+
+        final String topology = builder.build().describe().toString();
+
+        assertEquals(0, countOccurrencesInTopology(topology, "Sink: .*-repartition.*"));
+    }
+
+    @Test
+    public void shouldNotRepartitionWithMarkAsPartitionedFollowingMap() throws Exception {
```

Comment:
Not sure if we need to test this twice -- in the end, it does not matter which upstream operation sets `repartitionedRequired = true`.

However, we might want to add a test with a "broadcast" pattern:
```
// case 1: should get two downstream nodes and neither should have a repartition topic
KStream alreadyPartitionedStream = stream.map()markAsPartitioned();
alreadyPartitionedStream..groupByKey()
                .count()
                .toStream()
                .to(outputTopic);
alreadyPartitionedStream..groupByKey()
                .count()
                .toStream()
                .to(outputTopic2);
```
And also mix-and-match:
```
KStream needRepartition = stream.map():
map.markAsPartitioned()
                .groupByKey()
                .count()
                .toStream()
                .to(outputTopic); // should not insert repartition
// broadcast below should add "repartition step" and "branch topology"
map.groupByKey()
                .count()
                .toStream()
                .to(outputTopic2);
```

It might also be good to build a topology with "optimization enabled" to ensure we don't break anything inserting a new node.